sycl/doc/extensions/experimental/sycl_ext_oneapi_usm_raii.asciidoc

```diff
@@ -0,0 +1,169 @@
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== USM Deleter
+
+This extension adds `usm_deleter` helper that can be used with `std::unique_ptr`
+or `std::shared_ptr` to manage deallocation of USM memory. Unlike
+`usm_allocator` that doesn't work with `device` USM memory and can have
+unexpected implications when used through APIs like
+`std::make_shared`/`std::allocate_shared`, this new `usm_deleter` helper is
+generally applicable and doesn't cause surprises like unexpected concurrent
+access of the same memory page from both host/decive that could happen when
+using `usm_allocator` with shared pointers.
+
+[source, c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+class usm_deleter {
+public:
+  usm_deleter(const context &Ctx);
+  usm_deleter(const queue &q);
+
+  template <typename T> void operator()(T *Ptr) const;
+};
+} // sycl::ext::oneapi::experimental
+----
+
+=== Example usage
+
+A simple example of using this extension is below.
+
+```c++
+{
+  sycl::queue q;
+  using sycl::ext::oneapi::experimental::usm_deleter;
+  std::unique_ptr<int, usm_deleter> p{sycl::malloc_shared<int>(1, q), {q}};
+  q.single_task([=, p = p.get()]() { *p = 42; }).wait();
+}
+```
+
+=== Non-normative notes
+
+This helper doesn't introduce any synchronization points so the following would
+result in a race condition:
+
+```c++
+{
+  sycl::queue q;
+  using sycl::ext::oneapi::experimental::usm_deleter;
+  std::unique_ptr<int, usm_deleter> p{sycl::malloc_shared<int>(1, q), {q}};
+  q.single_task([=, p = p.get()]() { *p = 42; });
+  // No wait, memory might be free'ed before the kernel has finished execution.
+}
+```
+
+This isn't different from using smart pointers in multi-threaded environment
+outside SYCL though:
+
+```c++
+{
+  auto p = std::make_unique<...>();
+  std::thread t{[p = p.get()]() { use(p); }};
+  // No `t.join()`
+  // UB, memory is `delete`'ed before `t` finishes.
+}
+```
+
+== Issues/open questions/implementation details
+
+* link:https://en.cppreference.com/w/cpp/memory/shared_ptr/allocate_shared[`allocate_shared(const Alloc& alloc, Args&&... args )`]
+  seems to work with existing `usm_allocator`. However, it uses single
+  allocation for both the allocated object and auxilary data for
+  `std::shared_ptr`. Having that auxilary data residing in the same USM memory
+  might have unintended consequences depending on particular HW when copies are
+  created on host when the same memory page is being accessed on a device.
+* Providing `make_unique`/`make_shared`-like interface would require adding huge
+  number of APIs with unclear readability (e.g. "shared" USM vs "shared" smart
+  pointer), so not explored in this revision.
+* How should `usm_deleter` store `context`/`device`? Should it be another copy
+  following SYCL's common reference semantics and, essentially, extending their
```

Comment:
@intel/dpcpp-doc-reviewers , I think we need to answer this as part of the very first revision. What are your thoughts here?

Reply:
I'm not sure about this, but I want to point out that SYCL 2020 currently says that "[memory] must eventually be deallocated with `sycl::free` to avoid a memory leak".  I can't find anything to suggest that the `context` or `queue` destructor will call `sycl::free` on all of its USM pointers.

The current wording suggests to me that the only way to avoid a memory leak is to go with your first proposal, and prevent the `context` from being destructed until the pointer goes out of scope.

I'm curious to hear if @gmlueck agrees with my reading, though.  I might have missed something.

Reply:
Requiring the user to manually ensure the lifetime of the `context` object is at least as long as the smart pointer seems like a foot gun.  I think the smart pointer needs to make a copy internally.

Reply:
It should definitely bump the ref count on the shared_ptr of the context.  Do not circumvent that. Our "shutdown" procedure on Windows and Linux depend on the integrity of those ref counts especially for the "default context".